src/EventStore.Core/TransactionLog/Scavenging/Stages/StreamCalculator.cs

```diff
@@ -0,0 +1,122 @@
+
+			OriginalStreamHandle = originalStreamHandle;
+			OriginalStreamData = originalStreamData;
+		}
+
+		// State that doesn't change. scoped to the scavenge.
+		public IIndexReaderForCalculator<TStreamId> Index { get; }
+		public ScavengePoint ScavengePoint { get; }
+
+		// State that is scoped to the stream
+		public StreamHandle<TStreamId> OriginalStreamHandle { get; private set; }
+		private OriginalStreamData OriginalStreamData { get; set; }
+
+		// Returns NoStream (-1) if there are no events before the scavenge point.
+		// Caller must handle that
+		private long? _lastEventNumber;
+		public long LastEventNumber {
+			get {
+				if (!_lastEventNumber.HasValue) {
+					_lastEventNumber = Index.GetLastEventNumber(OriginalStreamHandle, ScavengePoint);
+				}
+
+				return _lastEventNumber.Value;
+			}
+		}
+
+		public DiscardPoint TruncateBeforeDiscardPoint =>
+			OriginalStreamData.TruncateBefore.HasValue
+				? DiscardPoint.DiscardBefore(OriginalStreamData.TruncateBefore.Value)
+				: DiscardPoint.KeepAll;
+
+		public DiscardPoint MaxCountDiscardPoint =>
+			// if LastEventNumber is NoStream (-1) this will always KeepAll as intended
+			OriginalStreamData.MaxCount.HasValue
+				? DiscardPoint.DiscardIncluding(LastEventNumber - OriginalStreamData.MaxCount.Value)
+				: DiscardPoint.KeepAll;
+
+		// accounts for tombstone, truncateBefore, maxCount.
+		private DiscardPoint? _truncateBeforeOrMaxCountDiscardPoint;
+		public DiscardPoint TruncateBeforeOrMaxCountDiscardPoint {
+			get {
+				if (!_truncateBeforeOrMaxCountDiscardPoint.HasValue) {
+					_truncateBeforeOrMaxCountDiscardPoint =
+						TruncateBeforeDiscardPoint.Or(MaxCountDiscardPoint);
+				}
+
+				return _truncateBeforeOrMaxCountDiscardPoint.Value;
+			}
+		}
+
+		public bool IsTombstoned => OriginalStreamData.IsTombstoned;
+
+		// We can discard the event when it is as old or older than the cutoff
+		public DateTime? CutoffTime => ScavengePoint.EffectiveNow - OriginalStreamData.MaxAge;
+
+		// Calculates whether this stream needs recalculating, assuming the metadata and istombstoned
+		// do not change (either of these updates will cause the calculator to reactivate it).
+		public CalculationStatus CalculateStatus() {
+			if (OriginalStreamData.IsTombstoned) {
+				// discard points will not move after this, BUT it cannot be deleted because we might
+				// run a scavenge with UnsafeHardDeletes in which case we will need to know this is
+				// tombstoned in order to discard the tombstone from the index.
+				return CalculationStatus.Archived;
+			}
+
+			if (OriginalStreamData.MaxAge.HasValue) {
+				//  because time will have passed so discard points might need moving
+				return CalculationStatus.Active;
+			}
+
+			if (OriginalStreamData.MaxCount.HasValue) {
+				// new evernts might have been added so discard point might need moving
+				// (unless the accumulator tracked when new events have been written per stream, but
+				// this would likely not be worth it.
```

Comment:
```suggestion
				// (unless the accumulator tracked when new events have been written per stream, but
				// this would likely not be worth it.)
```